source/common/http/async_client_impl.h

```diff
@@ -100,20 +100,37 @@ class AsyncStreamImpl : public virtual AsyncClient::Stream,
                   const AsyncClient::StreamOptions& options);
   ~AsyncStreamImpl() override {
     router_.onDestroy();
+    for (uint32_t i = 0; i < high_watermark_calls_; ++i) {
+      watermark_callbacks_->get().onDecoderFilterBelowWriteBufferLowWatermark();
```

Comment:
do we have a test covering this case?

Reply:
let's add `ENVOY_BUG` then?

I think it could happen if we watermarked, and got a close or reset from shadow upstream. Could you add a test for this scenario?

Reply:
sorry for being unclear, since this can happen in the case discussed above, then it's not an envoy bug if we have high wm calls > 0.

Reply:
+1 to extensive watermark testing :-)
I would expect for shared H2 connections you could end up destroying in high watermark state

Reply:
if it's not possible then remove the code handling it and leave it an envoy bug?  bonus points for a comment explaining it'll be done in clean up